common/FileSystem.cpp

```diff
@@ -271,6 +274,134 @@ std::wstring FileSystem::GetWin32Path(std::string_view str)
+			StringUtil::UTF8StringToWideString(FileSystem::GetWorkingDirectory()).c_str())))
+	{
+		return false;
+	}
+
+	// Set launch arguments.
+	if (!launch_arguments.empty())
+	{
+		if (check_failed(shell_link->SetArguments(StringUtil::UTF8StringToWideString(launch_arguments).c_str())))
+			return false;
+	}
+
+	// Set icon.
+	const std::string icon_path =
+		Path::Combine(Path::GetDirectory(FileSystem::GetProgramPath()), "resources\\icons\\AppIconLarge.ico");
+	if (check_failed(shell_link->SetIconLocation(StringUtil::UTF8StringToWideString(icon_path).c_str(), 0)))
+		return false;
+
+	// Set description (comment).
+	if (!description.empty())
+	{
+		if (check_failed(shell_link->SetDescription(StringUtil::UTF8StringToWideString(description).c_str())))
+			return false;
+	}
+
+	std::string shortcut_file_destination = "\\\\?\\";
+	// If no base directory ID specified, assume path suffix is an absolute path.
+	if (!IsEqualGUID(base_directory_ID, GUID_NULL))
+	{
+		PWSTR base_directory = nullptr;
+		if (check_failed(SHGetKnownFolderPath(base_directory_ID, 0, NULL, &base_directory)))
+			return false;
+		shortcut_file_destination += StringUtil::WideStringToUTF8String(base_directory) + "\\";
+		CoTaskMemFree(base_directory);
+	}
+
+	shortcut_file_destination += path_suffix;
+	const std::wstring shortcut_file_path = FileSystem::GetWin32Path(shortcut_file_destination);
+
+	Console.WriteLnFmt("Creating a shortcut '{}' with arguments '{}'",
+		StringUtil::WideStringToUTF8String(shortcut_file_path), launch_arguments);
+
+	// Use IPersistFile object to save shell link.
+	Microsoft::WRL::ComPtr<IPersistFile> persist_file;
+	if (check_failed(shell_link.As(&persist_file)))
+		return false;
+
+	// Save shortcut link to disk.
+	if (check_failed(persist_file->Save(shortcut_file_path.c_str(), TRUE)))
+		return false;
+
+	CoUninitialize();
+	return true;
+}
+#endif
+
+#if defined(__linux__) or defined(__FreeBSD__)
+bool FileSystem::CreateLinuxDesktopFile(const std::string_view destination, const std::string_view exec,
+	const std::string_view name, const std::string_view comment,
+	const std::string_view icon, const std::string_view categories,
+	const std::string_view terminal)
+{
+	// Create the .desktop file's contents.
+	const std::string file_content = fmt::format(
+		"[Desktop Entry]\n"
+		"Encoding=UTF-8\n"
+		"Version=1.0\n"
+		"Type=Application\n"
+		"Terminal={}\n"
+		"StartupWMClass=PCSX2\n"
+		"Exec={}\n"
+		"Name={}\n"
+		"Comment={}\n"
+		"Icon={}\n"
+		"Categories={}\n",
+		terminal, exec, name, comment, icon, categories);
+
+	// Write to .desktop file.
+	if (destination.empty() || !WriteStringToFile(destination.data(), file_content))
+	{
+		Console.ErrorFmt("Failed to create .desktop file at {}.", destination);
+		return false;
+	}
+
+	// Enable file execution.
+	if (chmod(destination.data(), S_IRWXU))
+		Console.WarningFmt("Failed to change file permissions for .desktop file: {} ({})", strerror(errno), errno);
```

Comment:
Would be nice to bubble the error message up to the caller here by passing a pointer to an Error object (see common/Error.h) as a parameter, and then filling it in with Error::SetString(error, "message..."). That way you can open a dialog box for it.

Reply:
Actually now I think about it code in common can't use the TRANSLATE macros, so that won't work.

Reply:
You could return an enum for the error, or if you don't like that you could move the code back to where it was.

Reply:
Looking at this again it's probably fine if you wanna leave it as-is.